### ISSUE TITLE
 WASAPI no sound with single channel

### ISSUE THREAD

**Issue by laiyonghao** [thestk/rtaudio#180]
Hello,

I compiled [playback sample](http://www.music.mcgill.ca/~gary/rtaudio/playback.html) with WASAPI, and it works fine. When I changed `parameters.nChannels = 2;` to `parameters.nChannels = 1;`, and `saw()` function codes accordingly,  no sound is out. But compiled with DirectSound, it is OK.

I need WASAPI for low latency, because my application playback audio stream over LAN.

Please Help.

> 
> compiler: VC2013
> OS: Windows 7 SP1
> RtAudio: 5.0.0

**Comment by MarcusTomlinson** [comment]
Interesting. I shall have a look.

**Comment by MarcusTomlinson** [comment]
Just to make sure, are you building RtAudio from the latest master?

I just pulled master and tried the sample as you described using a single channel, and it worked fine.

Here's the precise code I ran:

```
#include "RtAudio.h"
#include <iostream>
#include <cstdlib>
// Two-channel sawtooth wave generator.
int saw( void *outputBuffer, void *inputBuffer, unsigned int nBufferFrames,
         double streamTime, RtAudioStreamStatus status, void *userData )
{
  unsigned int i, j;
  double *buffer = (double *) outputBuffer;
  double *lastValues = (double *) userData;
  if ( status )
    std::cout << "Stream underflow detected!" << std::endl;
  // Write interleaved audio data.
  for ( i=0; i<nBufferFrames; i++ ) {
    for ( j=0; j<1; j++ ) {
      *buffer++ = lastValues[j];
      lastValues[j] += 0.005 * (j+1+(j*0.1));
      if ( lastValues[j] >= 1.0 ) lastValues[j] -= 2.0;
    }
  }
  return 0;
}
int main()
{
  RtAudio dac;
  if ( dac.getDeviceCount() < 1 ) {
    std::cout << "\nNo audio devices found!\n";
    exit( 0 );
  }
  RtAudio::StreamParameters parameters;
  parameters.deviceId = dac.getDefaultOutputDevice();
  parameters.nChannels = 1;
  parameters.firstChannel = 0;
  unsigned int sampleRate = 44100;
  unsigned int bufferFrames = 256; // 256 sample frames
  double data[2];
  try {
    dac.openStream( &parameters, NULL, RTAUDIO_FLOAT64,
                    sampleRate, &bufferFrames, &saw, (void *)&data );
    dac.startStream();
  }
  catch ( RtAudioError& e ) {
    e.printMessage();
    exit( 0 );
  }

  char input;
  std::cout << "\nPlaying ... press <enter> to quit.\n";
  std::cin.get( input );
  try {
    // Stop the stream
    dac.stopStream();
  }
  catch (RtAudioError& e) {
    e.printMessage();
  }
  if ( dac.isStreamOpen() ) dac.closeStream();
  return 0;
}
```

**Comment by MarcusTomlinson** [comment]
This is super weird... This works flawlessly when compiling with MinGW, but fails just as you describe with VC.

**Comment by MarcusTomlinson** [comment]
Alright I found the issue. Turns out that MinGW mallocs memory to all zeros, whereas VC++ does not. Really, the latter is more what I'd expect. All the former did was hide my mistake.

So, when the user channel count is less than the device channel count (as in this scenario), the unused channels were initialised with rubbish in them, which evidently WASAPI really didn't like.

Anyway, this is fixed now in: https://github.com/thestk/rtaudio/pull/183
